pkg/controller/rsm2/in_place_update_util.go

```diff
@@ -0,0 +1,345 @@
+	"github.com/apecloud/kubeblocks/pkg/dataprotection/utils"
+	viper "github.com/apecloud/kubeblocks/pkg/viperx"
+)
+
+type PodUpdatePolicy string
+
+const (
+	NoOpsPolicy         PodUpdatePolicy = "NoOps"
+	RecreatePolicy      PodUpdatePolicy = "Recreate"
+	InPlaceUpdatePolicy PodUpdatePolicy = "InPlaceUpdate"
+)
+
+func supportPodVerticalScaling() bool {
+	kubeVersion, err := utils.GetKubeVersion()
+	// if the Kubernetes version is unknown, assume pod vertical scaling is not supported.
+	if err != nil {
+		return false
+	}
+	if semver.Compare(kubeVersion, "v1.29") >= 0 {
+		return true
+	}
+
+	return utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling)
+}
+
+func filterInPlaceFields(src *corev1.PodTemplateSpec) *corev1.PodTemplateSpec {
+	template := src.DeepCopy()
+	// filter annotations
+	template.Annotations = nil
+	// filter labels
+	template.Labels = nil
+	// filter spec.containers[*].images & spec.initContainers[*].images
+	for i := range template.Spec.Containers {
+		template.Spec.Containers[i].Image = ""
+	}
+	for i := range template.Spec.InitContainers {
+		template.Spec.InitContainers[i].Image = ""
+	}
+	// filter spec.activeDeadlineSeconds
+	template.Spec.ActiveDeadlineSeconds = nil
+	// filter spec.tolerations
+	template.Spec.Tolerations = nil
+	// filter spec.containers[*].resources["cpu|memory"]
+	for i := range template.Spec.Containers {
+		delete(template.Spec.Containers[i].Resources.Requests, corev1.ResourceCPU)
+		delete(template.Spec.Containers[i].Resources.Requests, corev1.ResourceMemory)
+		delete(template.Spec.Containers[i].Resources.Limits, corev1.ResourceCPU)
+		delete(template.Spec.Containers[i].Resources.Limits, corev1.ResourceMemory)
+	}
+
+	return template
+}
+
+func copyRequestsNLimitsFields(container *corev1.Container) (corev1.ResourceList, corev1.ResourceList) {
+	requests := make(corev1.ResourceList)
+	limits := make(corev1.ResourceList)
+	if len(container.Resources.Requests) > 0 {
+		if requestCPU, ok := container.Resources.Requests[corev1.ResourceCPU]; ok {
+			requests[corev1.ResourceCPU] = requestCPU
+		}
+		if requestMemory, ok := container.Resources.Requests[corev1.ResourceMemory]; ok {
+			requests[corev1.ResourceMemory] = requestMemory
+		}
+	}
+	if len(container.Resources.Limits) > 0 {
+		if limitCPU, ok := container.Resources.Limits[corev1.ResourceCPU]; ok {
+			limits[corev1.ResourceCPU] = limitCPU
+		}
+		if limitMemory, ok := container.Resources.Limits[corev1.ResourceMemory]; ok {
+			limits[corev1.ResourceMemory] = limitMemory
+		}
+	}
+	return requests, limits
+}
+
+func mergeInPlaceFields(src, dst *corev1.Pod) {
+	mergeMap(&src.Annotations, &dst.Annotations)
+	mergeMap(&src.Labels, &dst.Labels)
+	dst.Spec.ActiveDeadlineSeconds = src.Spec.ActiveDeadlineSeconds
+	mergeList(&src.Spec.Tolerations, &dst.Spec.Tolerations, func(item corev1.Toleration) func(corev1.Toleration) bool {
+		return func(t corev1.Toleration) bool {
+			return false
```

Comment:
Why does it directly return false here? If the src Tolerations is not empty, won't the dst keep growing? And there's no deduplication.

Reply:
`Tolerations` is append-only according to the description of the [K8s docs](https://kubernetes.io/docs/concepts/workloads/pods/#pod-update-and-replacement)

Reply:
fixed: [054f0ac](https://github.com/apecloud/kubeblocks/pull/7000/commits/054f0ac0e5fac99a625df50c447c3128d47d48ad)